include/deal.II/fe/mapping_q_internal.h

```diff
@@ -1286,112 +1308,106 @@ namespace internal
-              {
-                double result[spacedim][dim];
+            if (update_flags & update_covariant_transformation)
+              for (unsigned int j = 0; j < n_lanes && i + j < n_points; ++j)
+                data.covariant[i + j] =
+                  data.contravariant[i + j].covariant_form();
 
-                // peel away part of sum to avoid zeroing the
-                // entries and adding for the first time
-                for (unsigned int i = 0; i < spacedim; ++i)
-                  for (unsigned int j = 0; j < dim; ++j)
-                    result[i][j] = data.derivative(point + data_set, 0)[j] *
-                                   data.mapping_support_points[0][i];
-                for (unsigned int k = 1; k < data.n_shape_functions; ++k)
-                  for (unsigned int i = 0; i < spacedim; ++i)
-                    for (unsigned int j = 0; j < dim; ++j)
-                      result[i][j] += data.derivative(point + data_set, k)[j] *
-                                      data.mapping_support_points[k][i];
-
-                // write result into contravariant data. for
-                // j=dim in the case dim<spacedim, there will
-                // never be any nonzero data that arrives in
-                // here, so it is ok anyway because it was
-                // initialized to zero at the initialization
-                for (unsigned int i = 0; i < spacedim; ++i)
-                  for (unsigned int j = 0; j < dim; ++j)
-                    data.contravariant[point][i][j] = result[i][j];
-              }
+            if (update_flags & update_inverse_jacobians)
+              for (unsigned int j = 0; j < n_lanes && i + j < n_points; ++j)
+                inverse_jacobians[i + j] = data.covariant[i + j].transpose();
           }
-
-      if (update_flags & update_covariant_transformation)
-        if (cell_similarity != CellSimilarity::translation)
+        else
           {
-            const unsigned int n_q_points = data.contravariant.size();
-            for (unsigned int point = 0; point < n_q_points; ++point)
+            const auto result =
+              internal::evaluate_tensor_product_value_and_gradient(
+                polynomials_1d,
+                support_points,
+                unit_points[i],
+                polynomial_degree == 1,
+                renumber_lexicographic_to_hierarchic);
+
+            if (update_flags & update_quadrature_points)
+              quadrature_points[i] = result.first;
+
+            if (cell_similarity == CellSimilarity::translation)
+              continue;
+
+            if (update_flags & update_contravariant_transformation)
               {
-                data.covariant[point] =
-                  (data.contravariant[point]).covariant_form();
+                DerivativeForm<1, spacedim, dim> jac_transposed = result.second;
+                data.contravariant[i] = jac_transposed.transpose();
```

Comment:
covariant_form?

Reply:
`covariant` is handled below. It should be correct the way it is.